libraries/shared/factories/event/repository.go

```diff
@@ -16,9 +16,146 @@
+	SchemaName     SchemaName
+	TableName      TableName
+	OrderedColumns []ColumnName // Defines the fields to insert, and in which order the table expects them
+	ColumnValues ColumnValues // Associated values for columns, restricted to []byte, bool, float64, int64, string, time.Time
+}
+
+// Stores memoised insertion queries to minimise computation
+var ModelToQuery = map[string]string{}
+
+// Get or create a DB insertion query for the model
+func GetMemoizedQuery(model InsertionModel) string {
+	// The schema and table name uniquely determines the insertion query, use that for memoization
+	queryKey := string(model.SchemaName) + string(model.TableName)
+	query, queryMemoized := ModelToQuery[queryKey]
+	if !queryMemoized {
+		query = GenerateInsertionQuery(model)
+		ModelToQuery[queryKey] = query
+	}
+	return query
+}
+
+// Creates an insertion query from an insertion model. Should be called through GetMemoizedQuery, so the query is not
+// generated on each call to Create.
+func GenerateInsertionQuery(model InsertionModel) string {
+	var valuePlaceholders []string
+	var updateOnConflict []string
+	for i := 0; i < len(model.OrderedColumns); i++ {
+		valuePlaceholder := fmt.Sprintf("$%d", 1+i)
+		valuePlaceholders = append(valuePlaceholders, valuePlaceholder)
+		updateOnConflict = append(updateOnConflict,
+			fmt.Sprintf("%s = %s", model.OrderedColumns[i], valuePlaceholder))
+	}
+
+	baseQuery := `INSERT INTO %v.%v (%v) VALUES(%v)
+		ON CONFLICT (header_id, log_id) DO UPDATE SET %v;`
+
+	return fmt.Sprintf(baseQuery,
+		model.SchemaName,
+		model.TableName,
+		joinOrderedColumns(model.OrderedColumns),
+		strings.Join(valuePlaceholders, ", "),
+		strings.Join(updateOnConflict, ", "))
+}
+
+/* Given an instance of InsertionModel, example below, generates an insertion query and persists to the DB.
+
+testModel = shared.InsertionModel{
+	SchemaName:     "maker"
+	TableName:      "testEvent",
+	OrderedColumns: []string{"header_id", "log_id", constants.IlkFK, constants.UrnFK, "variable1"},
+	ColumnValues: ColumnValues{
+		"header_id": 303
+		"log_id":   "1",
+		"variable1": "value1",
+		constants.IlkFK: 808,
+		constants.UrnFK: 909,
+	},
+}
+*/
+func Create(models []InsertionModel, db *postgres.DB) error {
+	if len(models) == 0 {
+		return fmt.Errorf("repository got empty model slice")
+	}
+
+	tx, dbErr := db.Beginx()
+	if dbErr != nil {
+		return dbErr
+	}
+
+	for _, model := range models {
+		// Maps can't be iterated over in a reliable manner, so we rely on OrderedColumns to define the order to insert
+		// tx.Exec is variadically typed in the args, so if we wrap in []interface{} we can apply them all automatically
+		var args []interface{}
+		for _, col := range model.OrderedColumns {
+			value := model.ColumnValues[col]
+			// Check whether or not PG can accept the type of value in the model
+			okPgValue := driver.IsValue(value)
+			if !okPgValue {
+				logrus.WithField("model", model).Errorf("PG cannot handle value of this type: %T", value)
```

Comment:
Is the idea that if this error is thrown, that the developer creating the converter will know that they need to have this field be a different type, that is accepted by postgres?

I wonder if there is a way to create an enum of accepted types? But then that couples us pretty tightly to go's postgres implementation, so maybe we don't want that? 🤔

Reply:
What you are describing is a typeclass, and afaik it's impossible in Go. Runtime checks with `reflect` et. al is as bleeding edge we can go :(

Reply:
But yes, what will ultimately happen if the converter implementer tries to insert types postgres can't handle, we'll log the issue :)